include/nbl/system/ISystem.h

```diff
@@ -0,0 +1,244 @@
+#ifndef __NBL_I_SYSTEM_H_INCLUDED__
+#define __NBL_I_SYSTEM_H_INCLUDED__
+
+#include <variant>
+#include "nbl/core/IReferenceCounted.h"
+#include "nbl/system/ISkippableAsyncQueueDispatcher.h"
+#include "nbl/system/IFileArchive.h"
+#include "nbl/system/IWriteFile.h"
+
+namespace nbl {
+namespace system
+{
+
+class ISystem final : public core::IReferenceCounted
+{
+public:
+    class ISystemCaller : core::IReferenceCounted
+    {
+    protected:
+        virtual ~ISystemCaller() = default;
+
+    public:
+        virtual core::smart_refctd_ptr<IReadFile> createReadFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual core::smart_refctd_ptr<IWriteFile> createWriteFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual uint32_t read(IReadFile* file, void* buffer, size_t offset, size_t size) = 0;
+        virtual uint32_t write(IWriteFile* file, const void* buffer, size_t offset, size_t size) = 0;
+    };
+
+protected:
+    struct SRequestType;
+    class CAsyncQueue;
+
+    class SFutureBase
+    {
+        friend ISystem;
+
+    protected:
+        bool valid_flag = false;
+        SRequestType* request;
+        CAsyncQueue* asyncQ;
+
+    public:
+        // SFuture is non-copyable and non-movable
+        SFutureBase(const SFutureBase&) = delete;
+
+        ~SFutureBase()
+        {
+            request->set_skip();
+        }
+
+        bool ready() const { return !request || request->ready; }
+        bool valid() const { return valid_flag; }
+
+        void wait()
+        {
+            asyncQ->waitForRequestCompletion(request[0]);
+        }
+    };
+
+    template <typename T>
+    class SFuture : public SFutureBase
+    {
+        alignas(T) uint8_t storage[sizeof(T)];
+
+    public:
+        ~SFuture()
+        {
+            if (valid_flag)
+                getStorage()->~T();
+        }
+
+        T* getStorage() { return reinterpret_cast<T*>(storage); }
+        T get()
+        {
+            wait();
+            assert(valid_flag);
+            valid_flag = false;
+            T* ptr = getStorage();
+            return std::move(ptr[0]);
```

Comment:
erm dont reset the valid flag

also no need to move the value out (if can just live here, padded by the atomic bool and request pointer) so just return a reference

If you want to be able to move T out, then declare a `move` method or something on top of the `get()`

Reply:
> erm dont reset the valid flag

im trying to mimic behaviour of std future,
also valid_flag is needed to denote if destructor have to call underlying object destructor or not (valid_flag=true means that an object was created on the storage and havent been moved....... oh ok i see my mistake, i should call destructor if it was moved as well)

Reply:
nah, if `get()` returns `T&` then user can just do `move(get())`

Reply:
> im trying to mimic behaviour of std future

maybe they didn't make the smartest design choice.

> also valid_flag is needed to denote if destructor have to call underlying object destructor or not (valid_flag=true means that an object was created on the storage and havent been moved....... oh ok i see my mistake, i should call destructor if it was moved as well)

Yeah pretty much, need to call the destructor all the time.